### ISSUE TITLE
False positives of react/prop-types in forwardRefs with 7.27.0

### ISSUE THREAD

**Issue by gunters63** [jsx-eslint/eslint-plugin-react#3140]
prop type checking seems to broken inside forwardRef with version 7.27.0, 7.26.1 still worked fine.

The error only happens for type intersection hierarchies (I didn't test interface extensions).

When you have types like this:

````
type ControlProps = {
  className?: string | undefined;
};

type NamedProps = {
  name: string;
};

type ControlPropsWithChildren = ControlProps & {
  children?: ReactNode;
};

type BaseButtonProps = ControlPropsWithChildren &
  NamedProps & {
    onClick?: (() => void) | undefined;
    onMouseDown?: (() => void) | undefined;
    onMouseUp?: (() => void) | undefined;
    disabled?: boolean | undefined;
    width?: number;
    type?: 'submit' | 'reset' | 'button' | undefined;
  };

````

The plugin will now falsely flag all properties in nested type intersections as missing in props validation, in this case name, className and children (see screenshot wiggly lines).

The problem does not happen when the component is not wrapped in forwardRef.
It also didn't happen with version 7.26.1

````
D:\...\BaseButton.tsx
  23:7  error  'name' is missing in props validation       react/prop-types
  24:7  error  'className' is missing in props validation  react/prop-types
  28:7  error  'children' is missing in props validation   react/prop-types
````

![image](https://user-images.githubusercontent.com/90346/141880939-bdd699a5-aae4-4194-b5b3-80729b0fed27.png)

Its likely that this commit: https://github.com/yannickcr/eslint-plugin-react/commit/4bc34995dbab597492c5081493d77755fd3501e8 triggered the problem

**Comment by ljharb** [comment]
cc @vedadeepta

**Comment by vaniyokk** [comment]
Same problem

**Comment by mataspetrikas** [comment]
I have noticed a potential cause/fix. 

The validation fails in the following case:

```
type InputTextProps = InputHTMLAttributes<HTMLInputElement> & {
  label: string;
  helperText?: string;
  error?: boolean;
};

export default forwardRef<HTMLInputElement, InputTextProps>(function InputText(
  props: InputTextProps,
  forwardedRef: Ref<HTMLInputElement>,
) {
  const { name, value, label, error = false, helperText, ...rest } = props;
  const [isFocused, setIsFocused] = useState<boolean>(false);
  const isActive = (typeof value === 'string' && value.length > 0) || isFocused;

  function handleFocus(event: FocusEvent<HTMLInputElement>) {
    setIsFocused(true);
    props.onFocus?.(event);
  }

  function handleBlur(event: FocusEvent<HTMLInputElement>) {
    if (!event.target.value) setIsFocused(false);
    props.onBlur?.(event);
  }

  function handleChange(event: ChangeEvent<HTMLInputElement>) {
    props.onChange?.(event);
  }

```

but it's all good when we change the name of the parameter to something else than `props`:

```


/**
 * A text input field with label and helper text.
 *
 * @example
 * <InputText label="Email" helperText="Please enter your email address" name="email" defaultValue="" error />
 */
export default forwardRef<HTMLInputElement, InputTextProps>(function InputText(
  componentProps: InputTextProps,
  forwardedRef: Ref<HTMLInputElement>,
) {
  const { name, value, label, error = false, helperText, ...rest } = componentProps;
```

Do we have a false positive that matches only the component props?

**Comment by gunters63** [comment]
Interesting.

On the other hand, the validation still fails when you destructure right away in the function parameters:

````
const BaseButton = forwardRef<HTMLButtonElement, BaseButtonProps>(
  (
    {
      name,
      className,
      onClick,
      onMouseDown,
      onMouseUp,
      children,
      disabled,
      width,
      type,
    }: BaseButtonProps,
    ref,
  ): JSX.Element => (
````

**Comment by nzdjb** [comment]
I've put together a reasonably minimal example that triggers this:

`import React, { forwardRef } from 'react'

interface D {
  x: string
  y: string
  z: string
}
type C = D
const A = forwardRef<HTMLDivElement, C>(({ x, y }, ref) => {
  return (
    <div ref={ref}>
      {x} {y}
    </div>
  )
})

A.displayName = 'A'
`

Issue seems to be that type is missing at https://github.com/yannickcr/eslint-plugin-react/blob/94826da35804b7ea1ae5adcd6ba4b8e049418cf1/lib/rules/prop-types.js#L107

**Comment by vedadeepta** [comment]
@gunters63 i'm not able to replicate. 

I pulled the latest and put your code almost exactly in the test case and it passed.

Here is the sample test case 
```
  valid: parsers.all([].concat(
    {

      code: `

        import React, { forwardRef } from 'react';
        type ControlProps = {
          className?: string | undefined;
        };

        type NamedProps = {
          name: string;
          test: number;
        };

        type ControlPropsWithChildren = ControlProps & {
          children?: ReactNode;
        };

        type ButtonProps = ControlPropsWithChildren &
          NamedProps & {
            onClick?: (() => void) | undefined;
            onMouseDown?: (() => void) | undefined;
            onMouseUp?: (() => void) | undefined;
            disabled?: boolean | undefined;
            width?: number;
            type?: 'submit' | 'reset' | 'button' | undefined;
        };

        const BaseButton = forwardRef<HTMLButtonElement, ButtonProps>((
          {
            name,
            className,
            onClick,
            onMouseDown,
            onMouseUp,
            children,
            disabled,
            width,
            type,
          },
          ref,
        ): JSX.Element => {
          return <span>{width}</span>
        })

      `,
      features: ['ts', 'no-babel'],

    },
  )),

  invalid: []
```

cc @ljharb

**Comment by ljharb** [comment]
Ideally we wouldnt have to explicitly support such things; the TS parser should have the type information we need already in the AST, thereby handling every possible type.

**Comment by gunters63** [comment]
This starts to get mysterious :)

If I inline the type definitions like this:

````
type NamedProps = {
  name: string;
};

type ControlProps = {
  className?: string | undefined;
};

type ControlPropsWithChildren = ControlProps & {
  children?: ReactNode;
};

type BaseButtonProps = ControlPropsWithChildren &
  NamedProps & {
    onClick?: (() => void) | undefined;
    onMouseDown?: (() => void) | undefined;
    onMouseUp?: (() => void) | undefined;
    disabled?: boolean | undefined;
    width?: number;
    type?: 'submit' | 'reset' | 'button' | undefined;
  };

const BaseButton = forwardRef<HTMLButtonElement, BaseButtonProps>(
  (
    {
   ....
````

the error goes away.

But my code originally does this:

````
import { ControlPropsWithChildren, NamedProps } from '$/components/BaseControl';

type BaseButtonProps = ControlPropsWithChildren &
  NamedProps & {
    onClick?: (() => void) | undefined;
    onMouseDown?: (() => void) | undefined;
    onMouseUp?: (() => void) | undefined;
    disabled?: boolean | undefined;
    width?: number;
    type?: 'submit' | 'reset' | 'button' | undefined;
  };

const BaseButton = forwardRef<HTMLButtonElement, BaseButtonProps>(
  (
    {
   ....
````

Contents of $/components/BaseControl:

````
import { ReactNode } from 'react';

export type ControlProps = {
  className?: string | undefined;
};

export type NamedProps = {
  name: string;
};

export type ControlPropsWithChildren = ControlProps & {
  children?: ReactNode;
};
````

If I import the base types like the code initially does the error comes back.

**Comment by vedadeepta** [comment]
@gunters63 ahh i see, if i understand correctly, currently we cannot parse types coming from external files.

@ljharb any thoughts on how to handle types coming from other files?

**Comment by gunters63** [comment]
@vedadeepta: 

I have many other components using those external types which do not use forwardRef. Those work fine.

**Comment by ljharb** [comment]
@vedadeepta the TS parser should provide all those types; we shouldn't have to figure them out ourselves.

**Comment by JameelKhan9** [comment]
Can also confirm exporting the type in the same file breaks only for forward refs

```
"@types/react": "^17.0.37",
"@typescript-eslint/eslint-plugin": "^5.6.0",
"@typescript-eslint/parser": "^5.6.0",
```
![image](https://user-images.githubusercontent.com/59159249/145155288-b95a3fe7-1ec2-4d10-9798-82203d4c275e.png)

The following is with the export:
![image](https://user-images.githubusercontent.com/59159249/145155420-544bc653-1a7c-4308-b57b-7cfb463a4ee3.png)

In saying this, I have no idea why exporting the type would break eslint from getting the proptype. 😕 

seems similar to the discussion happening at https://github.com/yannickcr/eslint-plugin-react/issues/3140

**Comment by vedadeepta** [comment]
@ljharb we're not getting the type in the AST. I checked the AST by creating a sample repo and importing types from other files.

Also its not just that the imported types don't work with `forwardRef<>` but they also don't work in general. we just don't throw an error for other cases.

I don't think this case can be handled without using the TS compiler API which can get the types from other files.

**Comment by gunters63** [comment]
@vedadeepta: As long as this is not implemented, wouldn't it be better to remove the special handling of forwardRef introduced in commit 4bc3499?

**Comment by vedadeepta** [comment]
Maybe, I'm not sure. This does not work for other React generic like `FC<>` etc, the only difference is that it fails silently

```
import { ControlProps, NamedProps } from './props';

type ButtonProps = ControlProps &
      NamedProps & {
        text: string;
    };
const Person2: React.FC<ButtonProps> = (props) => (
  <span>{props.somethingNotDef}</span> // <-- Does not throw an error but it should
);
```

So we will need to remove all ts handling since it cannot get types imported from other files

**Comment by ljharb** [comment]
In that case yes, we should remove it.

**Comment by ljharb** [comment]
Or at least, it should only warn when the entire type is defined in the same file?

**Comment by JDMathew** [comment]
An intermediate solution would be to wrap the FC export like so:

```
export interface IInputFieldProps extends InputHTMLAttributes<HTMLInputElement> {
...
}
const InputField = (props: IInputFieldProps, forwardedRef: Ref<HTMLInputElement>,) => {
return (
      <>
        <input {...props} ref={ref}></input>
      </>
    )
  });

export default forwardRef<HTMLInputElement, IInputFieldProps>(InputField);
```

**Comment by aakash-lohono** [comment]
hi,
another intermediate fix is to specify the argument types explicitly, like so

```tsx
import React, { ForwardedRef, forwardRef } from "react";

export interface HelloWorldProps {
  subject: string;
  className?: string;
}

const HelloWorld = forwardRef(
  function HelloWorldBase (
    props: HelloWorldProps,
    ref: ForwardRef<HtmlDivElement>,
  ) {
    const { subject, className } = props;

    return (
      <div 
        className={className}
        ref={ref}
      >
        Hello {subject}
      </div>
    );
  },
);

export default HelloWorld;
```

**Comment by gunters63** [comment]
> hi, another intermediate fix is to specify the argument types explicitly, like so
> 

Nice find! This workaround is low-effort and works well, I finally can unpin the package version again  :)

**Comment by ljharb** [comment]
@vedadeepta any chance you'd be able to work on a PR for this?

**Comment by mobily** [comment]
@ljharb I can take it if you and @vedadeepta don't mind :)

**Comment by ljharb** [comment]
That'd be great!

**Comment by vedadeepta** [comment]
thanks @mobily
